modules/ui/src/main/scala/lucuma/ui/forms/FormInputEV.scala

```diff
@@ -53,92 +55,135 @@ final case class FormInputEV[EV[_], A](
-  def valSet(s: String): Callback =
-    format.getOption(s).map(ev.set(value)).getOrEmpty
-  val onBlurC: InputEV.ChangeCallback[String]   =
-    (s: String) => format.getOption(s).map(onBlur).getOrEmpty
-  val onChangeC: InputEV.ChangeCallback[String] =
-    (s: String) => format.getOption(s).map(onChange).getOrEmpty
+
+  def valGet: String = ev.get(value).foldMap(validate.reverseGet)
+
+  def valSet: InputEV.ChangeCallback[A] = ev.set(value)
+
+  def onBlurC(onError: NonEmptyList[String] => Callback): InputEV.ChangeCallback[String] =
+    (s: String) =>
+      validate
+        .getValidated(s)
+        .fold(onError, onBlur)
 
   def withMods(mods: TagMod*): FormInputEV[EV, A] = copy(modifiers = modifiers ++ mods)
 }
 
 object FormInputEV {
   type Props[EV[_], A]   = FormInputEV[EV, A]
   type ChangeCallback[A] = A => Callback
-  type Scope[EV[_], A]   = RenderScope[Props[EV, A], State, Unit]
 
   @Lenses
-  final case class State(curValue: String, prevValue: String)
-
-  def onTextChange[EV[_], A]($ : Scope[EV, A]): ReactEventFromInput => Callback =
-    (e: ReactEventFromInput) => {
-      // Capture the value outside setState, react reuses the events
-      val v = e.target.value
-      // First update the internal state, then call the outside listener
-      $.setStateL(State.curValue)(v) *>
-        // Next 2 might not be called if the InputFormat returns None
-        $.props.valSet(v) *>
-        $.props.onChangeC(v)
+  final case class State(curValue: String, prevValue: String, errors: Option[NonEmptyList[String]])
+
+  class Backend[EV[_], A]($ : BackendScope[Props[EV, A], State]) {
+
+    def validate(
+      props:     Props[EV, A],
+      value:     String
+    )(
+      onValid:   A => Callback = _ => Callback.empty,
+      onInvalid: NonEmptyList[String] => Callback = _ => Callback.empty
+    ): Callback = {
+      val validated     = props.validate.getValidated(value)
+      val validChangeCB = props.onValidChange(validated.isValid)
+      val onValidCB     = validated match {
+        case Valid(a)   => onValid(a)
+        case Invalid(e) => onInvalid(e)
+      }
+      validChangeCB >> onValidCB
     }
 
-  def onBlur[EV[_], A]($ : Scope[EV, A], c: ChangeCallback[String]): Callback =
-    c($.state.curValue)
+    def onTextChange(props: Props[EV, A]): ReactEventFromInput => Callback =
+      (e: ReactEventFromInput) => {
+        // Capture the value outside setState, react reuses the events
+        val v = e.target.value
+        // First update the internal state, then call the outside listener
+        $.setStateL(State.curValue)(v) >> $.setStateL(State.errors)(none) >>
+          validate(props, v)(a =>
+            /*props.valSet(a) >>*/ props.onChange(a) // TODO Maybe we don't call valSet here???
+          )
+      }
+
+    def onBlur(props: Props[EV, A], state: State): Callback =
+      validate(props, state.curValue)(
+        a => props.valSet(a) >> props.onBlur(a),
```

Comment:
We could not due `valSet` and get rid of the `View` altogether and let the parent update its state with the callbacks only?

Reply:
If we want to model uniformly the behavior described above, in which the `InputFormEV` control will always perform validation on blur and set the model value if it's valid, then I think the current approach achieves that.

We could say that that is not the behavior that `InputFormEV` encapsulates and move the responsibility of the setting the model value to the parent (and even validation).

The idea here is to encapsulate logic that we are going to use in a lot of places without having to reimplement it repeatedly on callers. For cases where we don't want the `Input` control setting the model or validating, we can use `FormInput` directly.

Reply:
Right, I guess my question is on whether we should send the model to the parent via `View` or a `A => Callback`. It seems we are doing both now.

Reply:
Yes, we are doing both. 

As I said here https://github.com/gemini-hlsw/lucuma-ui/pull/130#discussion_r516036618, I'm more inclined to remove the `onChanged` callback.

Reply:
I agree